File: macos/Sources/ZharpCore/Terminal/Cell.swift
import Foundation

/// Style flags for a terminal cell.
public struct CellFlags: OptionSet, Hashable, Sendable {
    public let rawValue: UInt16
    public init(rawValue: UInt16) { self.rawValue = rawValue }

    public static let none            = CellFlags([])
    public static let bold            = CellFlags(rawValue: 1 << 0)
    public static let dim             = CellFlags(rawValue: 1 << 1)
    public static let italic          = CellFlags(rawValue: 1 << 2)
    public static let underline       = CellFlags(rawValue: 1 << 3)
    public static let blink           = CellFlags(rawValue: 1 << 4)
    public static let inverse         = CellFlags(rawValue: 1 << 5)
    public static let hidden          = CellFlags(rawValue: 1 << 6)
    public static let strikethrough   = CellFlags(rawValue: 1 << 7)
    public static let doubleUnderline = CellFlags(rawValue: 1 << 8)

    /// This cell is the right half of a double-width character.
    public static let wideTrailing    = CellFlags(rawValue: 1 << 9)
}

/// A terminal color: default (scheme-defined), an indexed palette entry (0-255),
/// or a 24-bit RGB value. Packed into a single UInt32.
public struct TerminalColor: Equatable, Hashable, Sendable {
    private static let kindDefault: UInt32 = 0 << 24
    private static let kindIndexed: UInt32 = 1 << 24
    private static let kindRgb: UInt32 = 2 << 24

    public let raw: UInt32

    private init(_ raw: UInt32) { self.raw = raw }

    public static let `default` = TerminalColor(kindDefault)

    public static func indexed(_ index: Int) -> TerminalColor {
        TerminalColor(kindIndexed | UInt32(index & 0xFF))
    }

    public static func rgb(_ r: Int, _ g: Int, _ b: Int) -> TerminalColor {
        TerminalColor(kindRgb | UInt32(((r & 0xFF) << 16) | ((g & 0xFF) << 8) | (b & 0xFF)))
    }

    public var isDefault: Bool { (raw >> 24) == 0 }
    public var isIndexed: Bool { (raw >> 24) == 1 }
    public var isRgb: Bool { (raw >> 24) == 2 }

    public var index: Int { Int(raw & 0xFF) }
    public var rgbValue: UInt32 { raw & 0xFFFFFF }
}

/// One character cell of the terminal grid.
public struct Cell: Equatable, Sendable {
    /// Unicode scalar value; 0 means an empty (blank) cell.
    public var rune: Int
    public var fg: TerminalColor
    public var bg: TerminalColor
    public var flags: CellFlags

    public init(rune: Int = 0,
                fg: TerminalColor = .default,
                bg: TerminalColor = .default,
                flags: CellFlags = .none) {
        self.rune = rune
        self.fg = fg
        self.bg = bg
        self.flags = flags
    }

    public var isBlank: Bool { rune == 0 || rune == 32 }
}

/// One row of cells plus line-level metadata.
public final class TerminalLine {
    public var cells: [Cell]

    /// True if this line soft-wrapped into the next (no hard newline).
    public var wrapped: Bool = false

    public init(cols: Int) {
        cells = Array(repeating: Cell(), count: max(0, cols))
    }

    public init(cols: Int, fill: Cell) {
        let blank = Cell()
        cells = Array(repeating: fill == blank ? blank : fill, count: max(0, cols))
    }

    public func fill(_ value: Cell) {
        for i in cells.indices { cells[i] = value }
        wrapped = false
    }

    /// Fills [from, to) with the given cell. Bounds are clamped.
    public func fillRange(_ from: Int, _ to: Int, _ value: Cell) {
        let lo = max(0, from)
        let hi = min(cells.count, to)
        if lo >= hi { return }
        for i in lo..<hi { cells[i] = value }
    }

    public func resize(cols: Int) {
        if cols == cells.count { return }
        if cols < cells.count {
            cells.removeLast(cells.count - cols)
        } else {
            cells.append(contentsOf: Array(repeating: Cell(), count: cols - cells.count))
        }
    }

    public func isEmpty() -> Bool {
        for c in cells where !c.isBlank || c.bg != .default {
            return false
        }
        return true
    }
}

File: windows/src/Zharp.Core/Terminal/Cell.cs
namespace Zharp.Core.Terminal;

/// <summary>Style flags for a terminal cell.</summary>
[Flags]
public enum CellFlags : ushort
{
    None = 0,
    Bold = 1 << 0,
    Dim = 1 << 1,
    Italic = 1 << 2,
    Underline = 1 << 3,
    Blink = 1 << 4,
    Inverse = 1 << 5,
    Hidden = 1 << 6,
    Strikethrough = 1 << 7,
    DoubleUnderline = 1 << 8,

    /// <summary>This cell is the right half of a double-width character.</summary>
    WideTrailing = 1 << 9,
}

/// <summary>
/// A terminal color: default (scheme-defined), an indexed palette entry (0-255),
/// or a 24-bit RGB value. Packed into a single uint.
/// </summary>
public readonly struct TerminalColor : IEquatable<TerminalColor>
{
    private const uint KindDefault = 0u << 24;
    private const uint KindIndexed = 1u << 24;
    private const uint KindRgb = 2u << 24;

    public readonly uint Raw;

    private TerminalColor(uint raw) => Raw = raw;

    public static readonly TerminalColor Default = new(KindDefault);

    public static TerminalColor Indexed(int index) =>
        new(KindIndexed | (uint)(index & 0xFF));

    public static TerminalColor Rgb(int r, int g, int b) =>
        new(KindRgb | (uint)((r & 0xFF) << 16 | (g & 0xFF) << 8 | (b & 0xFF)));

    public bool IsDefault => (Raw >> 24) == 0;
    public bool IsIndexed => (Raw >> 24) == 1;
    public bool IsRgb => (Raw >> 24) == 2;

    public int Index => (int)(Raw & 0xFF);
    public uint RgbValue => Raw & 0xFFFFFF;

    public bool Equals(TerminalColor other) => Raw == other.Raw;
    public override bool Equals(object? obj) => obj is TerminalColor c && c.Raw == Raw;
    public override int GetHashCode() => (int)Raw;
    public static bool operator ==(TerminalColor a, TerminalColor b) => a.Raw == b.Raw;
    public static bool operator !=(TerminalColor a, TerminalColor b) => a.Raw != b.Raw;
}

/// <summary>One character cell of the terminal grid.</summary>
public struct Cell
{
    /// <summary>Unicode code point; 0 means an empty (blank) cell.</summary>
    public int Rune;
    public TerminalColor Fg;
    public TerminalColor Bg;
    public CellFlags Flags;

    public readonly bool IsBlank => Rune == 0 || Rune == ' ';
}

/// <summary>One row of cells plus line-level metadata.</summary>
public sealed class TerminalLine
{
    public Cell[] Cells;

    /// <summary>True if this line soft-wrapped into the next (no hard newline).</summary>
    public bool Wrapped;

    public TerminalLine(int cols)
    {
        Cells = new Cell[cols];
    }

    public TerminalLine(int cols, in Cell fill)
    {
        Cells = new Cell[cols];
        if (fill.Rune != 0 || fill.Bg != TerminalColor.Default || fill.Fg != TerminalColor.Default || fill.Flags != CellFlags.None)
            Array.Fill(Cells, fill);
    }

    public void Fill(in Cell fill)
    {
        Array.Fill(Cells, fill);
        Wrapped = false;
    }

    /// <summary>Fills [from, to) with the given cell. Bounds are clamped.</summary>
    public void FillRange(int from, int to, in Cell fill)
    {
        from = Math.Max(0, from);
        to = Math.Min(Cells.Length, to);
        for (int i = from; i < to; i++)
            Cells[i] = fill;
    }

    public void Resize(int cols)
    {
        if (cols == Cells.Length)
            return;
        var next = new Cell[cols];
        Array.Copy(Cells, next, Math.Min(cols, Cells.Length));
        Cells = next;
    }

    public bool IsEmpty()
    {
        foreach (ref readonly var c in Cells.AsSpan())
        {
            if (!c.IsBlank || c.Bg != TerminalColor.Default)
                return false;
        }
        return true;
    }
}
